### DISCUSSION TITLE
Is Triggerer's blocked warning configurable?

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [apache/airflow#30151] [Q&A]
We occasionally get a message like this in the UI or logs:

```
ERROR - Triggerer's async thread was blocked for 0.24 seconds, likely by a badly-written trigger. Set PYTHONASYNCIODEBUG=1 to get more information on overrunning coroutines.
```

The thing is:

* PYTHONASYNCIODEBUG=1 is already set and nothing extra gets logged
* We are only using triggers included in airflow core or using Astronomer's third party package: https://github.com/astronomer/astronomer-providers
* We only ever see this "blocking" for a fraction of a second
* We peak at around ~100 triggers, most of them though sleep for a long time (e.g. 1 hour)
* Nothing is so critical that it can't wait a few seconds (in fact spawning up worker processes after a trigger completes seems to take this long anyway)
* We are running on a not too powerful VM, possible other VMs on the cluster are spiking at this time

So it would great to set the blocking warning to around 1 second, but I don't see any information in the docs to do this.

**Comment by potiuk** [comment]
No. it's not configurable. It is hard-coded to 200 ms.

**Reply by notatallshaw-gts** [reply]
Looking at the code I have found that the message is technically incorrect in one sense and misleading in another sense: https://github.com/apache/airflow/blob/2.5.2/airflow/jobs/triggerer_job.py#L325

This is the steps the function takes:

 * Records the time
 * Async sleeps for 0.1 seconds
 * Checks the difference between the current time and recorded time
 * Checks if the difference is greater than 0.2

However the message is technically incorrect because it only proves the blocking for at least 0.1 second, imagine this scenario

 * The monitoring function async sleeps for 0.1 seconds
 * A new task starts running 0.099 seconds later
 * The new task blocks for 0.15 seconds
 * The monitor function resumes and reports that it has been blocked for 0.24 seconds
 * But this is untrue, the monitor function was only blocked for 0.14 seconds

The message is also reads to me as misleading because it implies nothing got completed for the last 0.24 seconds, but in reality it only measures if the monitoring task was blocked, imagine the following scenario:

* The monitoring function async sleeps for 0.1 seconds
* 100 news tasks starts running 0.099 seconds later
* Each task blocks on average 0.01 seconds
* The monitoring task tries to wake up but it is at the back of the async queue
*  The monitor function eventually resumes and reports "Triggerer's async thread was blocked for 0.24 seconds..."

Again the correct blocking time was is 0.15 seconds, and in that time between the monitoring task ran many tasks ran and each task was only blocked on average for 0.05 seconds, I think a better message would be something like:

> ERROR - Triggerer's async monitor job was blocked for at least 0.14 seconds, ...

And then I think either both the frequency the monitor job runs and the threshold it alerts on should be configurable, or the threshold alert should be a multiple of the number of triggers running, e.g. 16ms per trigger.

Do you think PRs would be accepted for this?

**Reply by potiuk** [reply]
I think it is pretty arbitrary detection of things are wrong. If you can make PR to improve it, it is always welcome and the discussion on detaills is always better on a PR proposal where you will explain your idea and those who look at it will be able to comment on it.

So feel free to propose improvements

**Reply by notatallshaw-gts** [reply]
Agreed, I'll have a think on this and create a PR if I feel confident I have a better solution.
